moneriote/moneriote.py

```diff
@@ -0,0 +1,234 @@
+            log_err('bad height_discovery_method option', fatal=True)
+        self.md_height_discovery_method = md_height_discovery_method
+
+        # default Monero RPC port
+        self._m_rpc_port = 18089
+        self._blockchain_height = None
+
+        if not os.path.isfile(PATH_CACHE):
+            log_msg("Auto creating \'%s\'" % PATH_CACHE)
+            f = open(PATH_CACHE, 'a')
+            f.write('[]')
+            f.close()
+
+        self.monerod_check()
+
+    def main(self):
+        # get & set the current blockheight
+        height = self.monerod_get_height(method=self.md_height_discovery_method)
+        if not height or not isinstance(height, int):
+            log_err("Unable to fetch the current blockchain height")
+            return
+        self._blockchain_height = height
+
+        nodes = RpcNodeList()
+        nodes += RpcNodeList.cache_read(PATH_CACHE)  # from `cached_nodes.json`
+        if nodes:
+            nodes = self.scan(nodes, remove_invalid=True)
+
+        if len(nodes.nodes) <= 2:
+            peers = self.monerod_get_peers()  # from monerod
+            nodes += self.scan(peers, remove_invalid=True)
+
+        if nodes.nodes:
+            nodes.cache_write()
+
+        nodes.shuffle()
+        inserts = nodes.nodes[:self.dns_provider.max_records]
+        dns_nodes = self.dns_provider.get_records()
+
+        # insert new records
+        for node in inserts:
+            if node.address not in dns_nodes:
+                self.dns_provider.add_record(node)
+
+        # remove old records
+        for i, node in enumerate(dns_nodes):
+            if node.address not in nodes or i >= self.dns_provider.max_records:
+                self.dns_provider.delete_record(node)
+
+    def scan(self, nodes: RpcNodeList, remove_invalid=False):
+        """
+        Start threads checking known nodes to see if they're alive.
+        :param nodes:
+        :param remove_invalid: only return valid nodes when set to True
+        :return: valid nodes
+        """
+        if len(nodes) == 0:
+            return nodes
+
+        now = datetime.now()
+        log_msg('Scanning %d node(s) on port %d. This can take several minutes. Let it run.' % (
+            len(nodes), self._m_rpc_port))
+
+        pool = Pool(processes=CONFIG['concurrent_scans'])
+        nodes = RpcNodeList.from_list(pool.map(partial(RpcNode.is_valid, self._blockchain_height), nodes))
+        pool.close()
+        pool.join()
+
+        log_msg('Scanning %d node(s) done after %d seconds, found %d valid' % (
+            len(nodes), (datetime.now() - now).total_seconds(), len(nodes.valid(valid=True))))
+
+        if remove_invalid:
+            nodes = nodes.valid(valid=True)
+
+        return nodes
+
+    def monerod_check(self):
+        url = 'http://%s:%d' % (self.md_daemon_addr, self.md_daemon_port)
+
+        try:
+            resp = requests.get(url, timeout=2)
+            assert resp.status_code == 404
```

Comment:
@skftn , xmrchain ref was nicely fixed by last commit. And I tried both w/ and w/o auth. I found node with auth returns 401 here instead of 404 then shows not reachable.
But node w/o auth still gave 0 peer and didn't figure out yet:
```
[2018-09-27 21:42] Determining zone_id; looking for 'xmr-tw.org'
[2018-09-27 21:42] GET: https://api.cloudflare.com/client/v4/zones
[2018-09-27 21:42] Cloudflare zone_id 'xxxxxx' matched to 'xmr-tw.org'
[2018-09-27 21:42] GET: https://moneroblocks.info/api/get_stats/
[2018-09-27 21:42] GET: https://xmrchain.net/api/networkinfo
[2018-09-27 21:42] xmrchain height is 1670545
[2018-09-27 21:42] Reading 'C:\Users\walass\AppData\Local\Temp\moneriote-cache.json'
[2018-09-27 21:42] Loaded 0 nodes from 'C:\Users\walass\AppData\Local\Temp\moneriote-cache.json'
[2018-09-27 21:42] Spawning daemon; executing command 'print_pl'
[2018-09-27 21:42] Got peers from RPC: 0 node(s)
[2018-09-27 21:42] Fetching existing record(s) (testopennode.xmr-tw.org)
[2018-09-27 21:42] GET: https://api.cloudflare.com/client/v4/zones/xxxxxx/dns_records/
[2018-09-27 21:42] Sleeping for 600 seconds
```

Reply:
I tried `assert resp.status_code == 404 or resp.status_code == 401` and it works well on my auth node.

Reply:
I fixed it in my last commit. Thanks :)